File: frontend/src/api/dualPerspectives.ts
import { apiFetch } from './index'

export interface DualPerspective {
  id: number
  title: string
  source_type: string
  source_id: number | null
  source_date: string
  source_label: string
  user_view: string
  tuzhan_view: string
  tuzhan_view_origin: 'llm' | 'user'
  created_at: string
  updated_at: string
}

export interface DualAnchorCandidates {
  events: { id: number; label: string }[]
  diary: { id: number; label: string }[]
  goals: { id: number; label: string }[]
  artifacts: { id: number; label: string }[]
}

function parseItem(data: unknown): DualPerspective {
  const item = (data ?? {}) as Record<string, unknown>
  return {
    id: Number(item.id),
    title: String(item.title ?? ''),
    source_type: String(item.source_type ?? 'free'),
    source_id: item.source_id == null ? null : Number(item.source_id),
    source_date: String(item.source_date ?? ''),
    source_label: String(item.source_label ?? ''),
    user_view: String(item.user_view ?? ''),
    tuzhan_view: String(item.tuzhan_view ?? ''),
    tuzhan_view_origin: item.tuzhan_view_origin === 'llm' ? 'llm' : 'user',
    created_at: String(item.created_at ?? ''),
    updated_at: String(item.updated_at ?? ''),
  }
}

export async function listDualPerspectives(): Promise<DualPerspective[]> {
  const response = await apiFetch('/api/dual-perspectives')
  const data = await response.json()
  if (!response.ok || !data.ok) throw new Error(data.error || '双视角读取失败')
  return Array.isArray(data.items) ? data.items.map(parseItem) : []
}

export async function listDualAnchorCandidates(): Promise<DualAnchorCandidates> {
  const response = await apiFetch('/api/dual-perspectives/anchors')
  const data = await response.json()
  if (!response.ok || !data.ok) throw new Error(data.error || '候选经历读取失败')
  const pick = (value: unknown) =>
    Array.isArray(value)
      ? value.map((item) => {
          const row = (item ?? {}) as Record<string, unknown>
          return { id: Number(row.id), label: String(row.label ?? '') }
        })
      : []
  return {
    events: pick(data.events),
    diary: pick(data.diary),
    goals: pick(data.goals),
    artifacts: pick(data.artifacts),
  }
}

export async function createDualPerspective(payload: {
  title: string
  source_type: string
  source_id?: number | null
  user_view?: string
  tuzhan_view?: string
}): Promise<DualPerspective> {
  const response = await apiFetch('/api/dual-perspectives', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  const data = await response.json()
  if (!response.ok || !data.ok || !data.item) throw new Error(data.error || '这一页没有建成，再试一次')
  return parseItem(data.item)
}

export async function saveDualPerspectiveView(
  id: number,
  role: 'user' | 'tuzhan',
  content: string,
  origin: 'llm' | 'user' = 'user',
): Promise<DualPerspective> {
  const response = await apiFetch(`/api/dual-perspectives/${id}/view`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ role, content, origin }),
  })
  const data = await response.json()
  if (!response.ok || !data.ok || !data.item) throw new Error(data.error || '保存没有成功')
  return parseItem(data.item)
}

export async function generateTuzhanDraft(id: number): Promise<string> {
  const response = await apiFetch(`/api/dual-perspectives/${id}/tuzhan-draft`, { method: 'POST' })
  const data = await response.json()
  if (!response.ok || !data.ok) throw new Error(data.error || '草稿没有生成，稍后再试或代她写下这一段')
  return String(data.draft ?? '')
}

export async function deleteDualPerspective(id: number): Promise<void> {
  const response = await apiFetch(`/api/dual-perspectives/${id}`, { method: 'DELETE' })
  const data = await response.json()
  if (!response.ok || !data.ok) throw new Error(data.error || '删除没有成功')
}

File: backend/api/dual_perspectives.py
# -*- coding: utf-8 -*-
"""M8 双视角叙事 API：同一经历的双方解释页。feature flag 关闭时全部 403。"""
from __future__ import annotations

import asyncio

from fastapi import APIRouter
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from ..core import dual_perspectives
from ..core.config import config
from ..core.persona_profiles import active_user_id

router = APIRouter(prefix="/api/dual-perspectives", tags=["dual-perspectives"])


def _disabled() -> JSONResponse:
    return JSONResponse({"ok": False, "error": "双视角叙事未开启"}, status_code=403)


class CreateBody(BaseModel):
    title: str = Field()
    source_type: str = Field(default="free")
    source_id: int | None = Field(default=None)
    user_view: str = Field(default="")
    tuzhan_view: str = Field(default="")
    tuzhan_view_origin: str = Field(default="user")


class ViewBody(BaseModel):
    role: str = Field(pattern="^(user|tuzhan)$")
    content: str = Field(default="")
    origin: str = Field(default="user", pattern="^(llm|user)$")


@router.get("")
async def api_list():
    if not config.dual_perspectives_enabled:
        return _disabled()
    return {"ok": True, "items": await asyncio.to_thread(dual_perspectives.list_perspectives, active_user_id())}


@router.get("/anchors")
async def api_anchor_candidates():
    if not config.dual_perspectives_enabled:
        return _disabled()
    return {"ok": True, **(await asyncio.to_thread(dual_perspectives.anchor_candidates, active_user_id()))}


@router.post("")
async def api_create(body: CreateBody):
    if not config.dual_perspectives_enabled:
        return _disabled()
    try:
        item = await asyncio.to_thread(
            dual_perspectives.create_perspective,
            active_user_id(),
            body.title,
            source_type=body.source_type,
            source_id=body.source_id,
            user_view=body.user_view,
            tuzhan_view=body.tuzhan_view,
            tuzhan_view_origin=body.tuzhan_view_origin,
        )
    except dual_perspectives.DualPerspectiveError as exc:
        return JSONResponse({"ok": False, "error": str(exc)}, status_code=400)
    return {"ok": True, "item": item}


@router.put("/{perspective_id}/view")
async def api_set_view(perspective_id: int, body: ViewBody):
    if not config.dual_perspectives_enabled:
        return _disabled()
    try:
        item = await asyncio.to_thread(
            dual_perspectives.set_view,
            active_user_id(),
            perspective_id,
            body.role,
            body.content,
            origin=body.origin,
        )
    except dual_perspectives.DualPerspectiveError as exc:
        return JSONResponse({"ok": False, "error": str(exc)}, status_code=400)
    return {"ok": True, "item": item}


@router.post("/{perspective_id}/tuzhan-draft")
async def api_tuzhan_draft(perspective_id: int):
    if not config.dual_perspectives_enabled:
        return _disabled()
    try:
        return await dual_perspectives.generate_tuzhan_draft(active_user_id(), perspective_id)
    except dual_perspectives.DualPerspectiveError as exc:
        return JSONResponse({"ok": False, "error": str(exc)}, status_code=400)


@router.delete("/{perspective_id}")
async def api_delete(perspective_id: int):
    if not config.dual_perspectives_enabled:
        return _disabled()
    deleted = await asyncio.to_thread(
        dual_perspectives.delete_perspective, active_user_id(), perspective_id
    )
    if not deleted:
        return JSONResponse({"ok": False, "error": "这一页不存在"}, status_code=404)
    return {"ok": True}
